libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java

```diff
@@ -305,6 +349,71 @@ static URI rootURI(URL url) {
         }
     }
 
+    static String removeMRJARPrefix(Path path) {
+        assert path.startsWith(MRJAR_VERSION_PREFIX) : path;
+        int c = path.getNameCount();
+        assert c >= 3 : path.getNameCount();
+        return c == 3 ? "" : path.subpath(3, path.getNameCount()).toString();
+    }
+
+    static int getMRVersionFromPrefix(Path path) {
+        assert path.startsWith(MRJAR_VERSION_PREFIX) : path;
+        int c = path.getNameCount();
+        assert c >= 3 : path.getNameCount();
+        return Integer.valueOf(path.subpath(2, 3).toString());
+    }
+
+    static String toPackageName(String name) {
+        assert name.endsWith("/") == false;
+        int index = name.lastIndexOf('/');
+        if (index != -1) {
+            return name.substring(0, index).replace('/', '.');
+        }
+        return name;
+    }
+
+    /** Tuple of package and package-to-release-version map. */
+    record ScanResult(Set<String> packages, Map<String, List<Integer>> pkgVersions) {}
+
+    static ScanResult scanPackages(Path dir) {
+        String separator = dir.getFileSystem().getSeparator();
+        Set<String> pkgs = new HashSet<>();
+        Map<String, List<Integer>> pkgVersions = new HashMap<>();
+        try {
+            try (var paths = Files.find(dir, Integer.MAX_VALUE, ((path, attrs) -> attrs.isRegularFile()))) {
+                paths.map(path -> dir.relativize(path)).filter(path -> path.getFileName().toString().endsWith(".class")).forEach(path -> {
```

Comment:
Resources can be versioned too, but I'm not sure that anyone is using that. I deliberately narrowed this change to classes/packages only, and left resources as is (scanning all possible versions), since resources are less of an issue. Also, from a modules perspective, pkgs and directories with non-class files are two different things. I have yet to find a good compromise to allow for both.

Reply:
Oh, I see - "For non MR packages we can just iterate directories" - Yes, we could.  The set of packages is quite small, and the pkg -> jarMeta is avoids poking about in search. Also, given that we load these classes within modules we can, and should, not delegate for any classes in packages known to this loader.